File: test/H2O.jl
include("../src/HartreeFock.jl")

using .HartreeFock
using BenchmarkTools

function H2O(r::Real, θ::Real)
    Molecule("H2O",
        [
            Atom("O", vec3(0, 0, 0)),
            Atom("H", vec3(r*cosd(θ/2),  r*sind(θ/2), 0)),
            Atom("H", vec3(r*cosd(θ/2), -r*sind(θ/2), 0))
        ]
    )
end

function H2O_pes(r_range::AbstractArray, θ_range::AbstractArray)
    Nr = length(r_range)
    Nt = length(θ_range)
    total_energy = zeros(Float64, Nr, Nt)
    Threads.@threads for i in 1:Nr
    # for i in 1:Nr
        r = r_range[i] / a0
        for j in  1:Nt
            θ = θ_range[j]
            e = hartree_fock(H2O(r, θ))
            total_energy[i, j] = e
        end
    end
    i, j = argmin(total_energy).I
    println("minimal energy of H2O is $(total_energy[i,j]*hartree)eV, with the bond length = $(r_range[i])Å, bond angle = $(θ_range[j])ᵒ")
end

# display(@benchmark H2O_pes(0.9:0.01:1.1, 85:115))
H2O_pes(0.9:0.01:1.1, 85:115)

File: test/H2O.py
# compare H2O result with pyscf

from time import time_ns
import numpy as np
from pyscf import gto
from pyscf import scf

a0 = 0.529_177_210_903
hartree = 27.211386245988

def H2O(r:float, theta:float):
    theta = theta * np.pi / 180
    return gto.M(
        atom = [
            ['O', [0, 0, 0]],
            ['H', [r * np.cos(theta/2), r * np.sin(theta/2), 0]],
            ['H', [r * np.cos(theta/2), -r * np.sin(theta/2), 0]]],
        basis="sto-3g",
        verbose=0
    )

def H2O_pes(r_range, theta_range):
    Nr = len(r_range)
    Nt = len(theta_range)
    Emat = np.zeros((Nr, Nt))
    for i, r in enumerate(r_range):
        for j, theta in enumerate(theta_range):
            hf = scf.RHF(H2O(r, theta))
            hf.chkfile = None
            Emat[i,j] = hf.kernel()
    pos = np.argmin(Emat)
    i, j = pos // Nt, pos % Nt
    print(f"minimal energy is {Emat[i,j]*hartree}eV, with the bond length = {r_range[i]:.2f}Å, bond angle = {theta_range[j]:.1f}ᵒ")

if __name__ == "__main__":
    start = time_ns()
    H2O_pes(np.linspace(0.9, 1.1, 21), np.linspace(85, 115, 31))
    stop = time_ns()
    print("time used:", (stop-start)/1e9)
